portfolio/src/main/java/com/google/sps/servlets/DataServlet.java

```diff
@@ -54,13 +56,21 @@ protected void doPost(HttpServletRequest req, HttpServletResponse resp) throws S
 
   @Override
   public void doGet(HttpServletRequest request, HttpServletResponse response) throws IOException {
-    Query query = new Query("comment");
-    PreparedQuery results = datastore.prepare(query);
-    ArrayList<String> comments = new ArrayList<>();
-    for (Entity entity : results.asIterable()) {
-      comments.add((String)entity.getProperty("value"));
+    String maxCommentsString = request.getParameter("max_comments");
```

Comment:
What might be more reasonable behavior in the event the query does not supply a max_comments key, or supplies a malformed one?

Reply:
Will change the servlet to return a error with return code 400.

Reply:
Returning an error is not un-reasonable, so I would be fine with returning an error. However, what I'd like to consider more broadly is the distinction between recoverable and non-recoverable errors, as well as thinking in terms of how client's might like to use an API.

Let's say you're a client making a request to a data API like this - what behavior would you expect if you don't include the max_comments value? Consider the similar API PreparedQuery exposes, what choices have devs made for when clients elide a similar parameter? As a general rule of thumb, you want APIs to be strict when enforcing correctness, and loose in terms useability (as useable as possible, awkward phrasing aside). Does not passing in the max_comments value affect the ability for your API to be correct, and does it expose any surprising behavior to the user?

Yet another angle on the question: What's best for the user? From my perspective an average user clearly wouldn't care what the exact value of max_comments is, and asking them to care seems unreasonable in the majority of cases. So then someone else has to care - and that leaves only the client or the server. The question is then, who is better placed to make the decision? When making a decision, the client has user specific information, and client information, but not server information. The server has user specific information, client specific information, and server information, so the server has the most information if a decision needs to be made. Arguably, the server is thus the most natural place to make a decision on what should happen if the user doesn't care about the value of max_comments.

If you think through these cases, and believe that the responsibility should be with the client (ie, forcing the client to include max_comments), then I'm fine with that - I want to see your reasoning however :)